src/mlpack/tests/ann/layer/sum_reduce.cpp

```diff
@@ -0,0 +1,113 @@
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#include <mlpack/core.hpp>
+#include <mlpack/methods/ann.hpp>
+
+#include "../../test_catch_tools.hpp"
+#include "../../catch.hpp"
+#include "../../serialization.hpp"
+#include "../ann_test_tools.hpp"
+
+using namespace mlpack;
+
+/**
+ * SumReduce input dimensions not set.
+ */
+TEST_CASE("SumReduceInputDimensionsNotSet", "[ANNLayerTest]")
+{
+  SumReduce layer(0);
+  REQUIRE_THROWS_AS(layer.ComputeOutputDimensions(), std::logic_error);
+}
+
+/**
+ * SumReduce axis is out of bounds.
+ */
+TEST_CASE("SumReduceAxisOutOfBounds", "[ANNLayerTest]")
+{
+  SumReduce layer(-1);
+  layer.InputDimensions() = { 2, 2 };
+  REQUIRE_THROWS_AS(layer.ComputeOutputDimensions(), std::logic_error);
+}
+
+/**
+ * Set keep dimension to false.
+ */
+TEST_CASE("SumReduceKeepDimensionsFalse", "[ANNLayerTest]")
+{
+  SumReduce layer(1, false);
+  layer.InputDimensions() = { 2, 8 };
+  layer.ComputeOutputDimensions();
+
+  std::vector<size_t> correctOutputDimensions = { 2 };
+  REQUIRE(layer.OutputDimensions() == correctOutputDimensions);
+}
+
+/**
+ * Set keep dimension to true.
+ */
+TEST_CASE("SumReduceKeepDimensionsTrue", "[ANNLayerTest]")
+{
+  SumReduce layer(1, true);
+  layer.InputDimensions() = { 3, 2, 2 };
+  layer.ComputeOutputDimensions();
+
+  std::vector<size_t> correctOutputDimensions = { 3, 1, 2 };
+  REQUIRE(layer.OutputDimensions() == correctOutputDimensions);
+}
+
+/**
+ * Test sum reduce forward pass.
+ */
+TEST_CASE("SumReduceForwardPass", "[ANNLayerTest]")
+{
+  size_t batchSize = 3;
+  arma::mat input = arma::regspace(0, 12 * batchSize - 1);
+  arma::mat output;
+
+  arma::mat expectedOutput = arma::mat(
+    {6, 8, 10, 12, 14, 16, 30, 32, 34, 36, 38, 40, 54, 56, 58, 60, 62, 64});
+  expectedOutput.reshape(6, batchSize);
+
+  SumReduce layer(2, false);
+  layer.InputDimensions() = { 3, 2, 2 };
+  layer.ComputeOutputDimensions();
+
+  input.reshape(12, batchSize);
+  layer.Forward(input, output);
+  CheckMatrices(output, expectedOutput);
+}
+
+/**
+ * Test sum reduce backward pass.
+ */
+TEST_CASE("SumReduceBackwardPass", "[ANNLayerTest]")
+{
+  size_t batchSize = 3;
+  arma::mat input, output;
+  input.set_size(12, batchSize);
+
+  arma::mat gy = arma::regspace(0, 6 * batchSize - 1);
+  gy.reshape(6, batchSize);
+  arma::mat g;
+
+  SumReduce layer(2, false);
+  layer.InputDimensions() = { 3, 2, 2 };
+  layer.ComputeOutputDimensions();
+
+  layer.Backward(input, output, gy, g);
+
+  arma::mat expectedG = arma::mat(
+    { 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
+      6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 12, 13, 14, 15, 16, 17 });
+  expectedG.reshape(12, batchSize);
+  CheckMatrices(g, expectedG);
+}
```

Comment:
Want to add a simple Jacobian test?  I think this could work:

```
/**
 * JacobianTest for LogSoftMax layer
 */
TEST_CASE("SumReduceJacobianTest", "[ANNLayerTest]")
{
  for (size_t i = 0; i < 5; ++i)
  {
    arma::mat input(3 * 4 * 5, 1, fill::randu);

    SumReduce module(2, false);
    module.InputDimensions() = { 3, 4, 5 };
    module.ComputeOutputDimensions();

    double error = JacobianTest(module, input);
    REQUIRE(error <= 1e-5);
  }
}
```

I adapted that from the `LogSoftmax` test, but I didn't try to compile or run it.  I think it should work, everything looks right to me.